main/javascript-modifications/ses-guide.md

```diff
@@ -0,0 +1,506 @@
+intrinsics, omitting `Date.now()` and `Math.random()`. These are left out
+since they can be covert communication channels between programs. 
+
+However, a compartment may be expressly given access to these objects through
+the compartment constructor's first argument or by assigning them to the 
+compartment's `globalThis` after construction.
+```js
+const powerfulCompartment = new Compartment({ Math });
+powerfulCompartment.globalThis.Date = Date;
+```
+
+When a new `Compartment` object is created, you must decide if it supports object-capability security.
+If it does, run `harden(compartment.globalThis)` on it before loading any untrusted code into it.
+
+A single compartment can run a JavaScript program in the locked-down SES environment.
+However, most interesting programs have multiple modules. So, each compartment also has
+its own module system. The next SES release will include support for ECMAScript modules,
+**tyg todo: Has this happened already? If not, is there an ETA?**
+a relatively new system supported by many browsers, and officially released in Node.js 14.
+
+Compartments can be linked, so one compartment can export a module that another compartment
+imports. Each compartment may have its own rules for how to resolve import specifiers and 
+how to locate and retrieve modules. In this example, we use the compartment constructor to 
+create two compartments: one for the application and another for its dependency.
+
+**tyg todo: Will people understand this code or do we need to explain any/all methods in it
+in more detail**
+
+```js
+const dependency = new Compartment({}, {}, {
+  resolveHook: (moduleSpecifier, moduleReferrer) =>
+    resolve(moduleSpecifier, moduleReferrer),
+  importHook: async moduleSpecifier => {
+    const moduleLocation = locate(moduleSpecifier);
+    const moduleText = await retrieve(moduleLocation);
+    return new ModuleStaticRecord(moduleText, moduleLocation);
+  },
+});
+const application = new Compartment({}, {
+  'dependency': dependency.module('./main.js'),
+}, {
+  resolveHook,
+  importHook,
+});
+```
+
+Compartments greatly simplify using SES for large applications. We will build tools that use 
+compartments to load and bundle existing libraries and then execute them with a compartment
+for each package. **tyg todo Has this happened? If so, where/what are the tools? If not, any ETA?**
+This will provide a seamless experience as simple as <script src=“app.js”> or node app.js, but 
+with the safety of SES.
+  
+**tyg todo: My problem with this section is that while I think I understand what compartments are
+and what they do, I'm not clear on just how/when they are used. This (and the SES reference) could
+use what amounts to a paragraph or two like "When writing a contract, you use compartments like this. 
+This is when you want to create a compartment. This is when you don't want to. Do you ever destroy a
+compartment? If so, how?**
```

Comment:
Compartments provide a low-level loader API for JavaScript modules. Your code might run in compartments, but they are an implementation detail of tools and runtimes.

The Agoric runtime uses compartments to isolate contracts within a Vat. MetaMask’s LavaMoat uses a Compartment for every module, to create boundaries between application code and third-party dependencies.

The lifetime of a compartment is bounded by garbage collection and the lifetime of the realm that contains them. There is no ceremony for tearing down a compartment.

Reply:
Changed last sentence to "You will not ever have to tear
down or delete one. " Is that OK?

Reply:
Is each compartment automatically a separate vat?

Reply:
Vats use compartments and can use multiple compartments. Compartments are not vats.

Reply:
Please let me know if you need further clarifications to settle this section.